framework/math/quadratures/angular/curvilinear_product_quadrature.cc

```diff
@@ -0,0 +1,417 @@
+    {
+      azimu_quad.weights.emplace(azimu_quad.weights.end(), 0);
+      azimu_quad.qpoints.emplace(azimu_quad.qpoints.end(), azimu_quad_span.second);
+    }
+  }
+
+  // Product quadrature initialization
+  // Compute weights, abscissae $(\varphi, \vartheta)$ and direction vectors
+  // $\omega_{pq} := (\mu_{pq}, \xi_{p}, \eta_{pq})$
+  weights.clear();
+  abscissae.clear();
+  omegas.clear();
+  for (size_t p = 0; p < azimu_quad_vec.size(); ++p)
+  {
+    const auto pol_wei = polar_quad.weights[p];
+    const auto pol_abs = polar_quad.qpoints[p][0];
+    const auto pol_com = std::sqrt(1 - pol_abs * pol_abs);
+
+    for (size_t q = 0; q < azimu_quad_vec[p].weights.size(); ++q)
+    {
+      const auto& azimu_quad = azimu_quad_vec[p];
+
+      const auto azi_wei = azimu_quad.weights[q];
+      const auto azi_abs = azimu_quad.qpoints[q][0];
+      const auto azi_com = std::sqrt(1 - azi_abs * azi_abs);
+
+      const auto weight = pol_wei * azi_wei;
+      const auto abscissa = QuadraturePointPhiTheta(std::acos(azi_abs), std::acos(pol_abs));
+      const auto omega = Vector3(pol_com * azi_abs, pol_abs, pol_com * azi_com);
+
+      weights.emplace_back(weight);
+      abscissae.emplace_back(abscissa);
+      omegas.emplace_back(omega);
+    }
+  }
+  weights.shrink_to_fit();
+  abscissae.shrink_to_fit();
+  omegas.shrink_to_fit();
+
+  // Map of direction indices
+  unsigned int ind0 = 0;
+  map_directions_.clear();
+  for (size_t p = 0; p < azimu_quad_vec.size(); ++p)
+  {
+    std::vector<unsigned int> vec_directions_p;
+    for (size_t q = 0; q < azimu_quad_vec[p].weights.size(); ++q)
+      vec_directions_p.emplace_back(ind0 + q);
+    map_directions_.emplace(p, vec_directions_p);
+    ind0 += azimu_quad_vec[p].weights.size();
+  }
+
+  // Curvilinear product quadrature
+  // Compute additional parametrising factors
+  InitializeParameters();
+
+  // Print
+  if (verbose)
+  {
+    log.Log() << "map_directions" << std::endl;
+    for (const auto& dir : map_directions_)
+    {
+      log.Log() << "polar level " << dir.first << " : ";
+      for (const auto& q : dir.second)
+        log.Log() << q << ", ";
+      log.Log() << std::endl;
+    }
+    log.Log() << "curvilinear product quadrature : cylindrical" << std::endl;
+    for (size_t k = 0; k < weights.size(); ++k)
+      log.Log() << "angle index " << k << ": weight = " << weights[k] << ", (phi, theta) = ("
+                << abscissae[k].phi << ", " << abscissae[k].theta << ")"
+                << ", omega = " << omegas[k].PrintStr()
+                << ", fac_diamond_difference = " << fac_diamond_difference_[k]
+                << ", fac_streaming_operator = " << fac_streaming_operator_[k] << std::endl;
+    const auto sum_weights = std::accumulate(weights.begin(), weights.end(), 0.0);
+    log.Log() << "sum(weights) = " << sum_weights << std::endl;
+  }
+}
+
+void
+GLCProductQuadrature2DRZ ::InitializeParameters()
```

Comment:
Interesting that the space between `GLCProductQuadrature2DRZ` and `::` was not found by clang-format check.

Reply:
I'm surprised by that too!  Fixed.